### PR TITLE
Feature/upgradeable contract hr

### PR DESCRIPTION
**Pull request by hermannleboss**
# Description

- [ ] Inprove performance of the smart contrac
- [ ] Add more security
- [ ] Make the smart contract upgradable. 

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

* **Please check if the PR fulfills these requirements**

- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: contract/contracts/Tips.sol [hunk 1]
```diff
@@ -18,19 +21,40 @@ contract Tips {
...
+        require(_teamMembersAddresses.length > 0, "Must have at least one team member");
+        require(
+            pushLimit >= _teamMembersAddresses.length,
+            "You have too much team members"
+        );
+        uint totalAmount = msg.value + remainder;
+        uint256 amountPerAddress = totalAmount / _teamMembersAddresses.length;
+        
+        // Update the remainder
+        remainder= totalAmount % _teamMembersAddresses.length;
```
Comment (farrellh1) [inline contract/contracts/Tips.sol:35]: i still dont understand what is it for, can you explain it @hermannleboss ?

File: contract/contracts/Tips.sol [hunk 2]
```diff
@@ -39,22 +63,29 @@ contract Tips {
...
+        uint256 senderBalance = balance[msg.sender];
+        require(senderBalance > 0, "No tips to withdraw.");
 
-        bool tipSuccess = payable(msg.sender).send(tipEarned);
-        require(tipSuccess, "Failed to withdraw tips.");
+        (bool sent,) = msg.sender.call{value: senderBalance}("");
+        require(sent, "Failed to withdraw tips.");
 
         balance[msg.sender] = 0;
+        require(balances[msg.sender] == 0, "Failed to zero out balance");
```
Comment (farrellh1) [inline contract/contracts/Tips.sol:75]: will this error happen? in my opinion i think it wont happen

Comment (hermannleboss) [reply contract/contracts/Tips.sol:75]: It should'nt but it's a security.

File: contract/contracts/Tips.sol (hunk 1 above)
Comment (farrellh1) [reply contract/contracts/Tips.sol:44]: i just realize that solidity can have decimal number, still have to think about better solution 🤔

File: contract/contracts/Tips.sol [hunk 3]
```diff
@@ -18,45 +25,95 @@ contract Tips {
...
 
-        bool tipSuccess = payable(msg.sender).send(tipEarned);
-        require(tipSuccess, "Failed to withdraw tips.");
+        (bool sent,) = msg.sender.call{value: senderBalance}("");
+        require(sent, "Failed to withdraw tips.");
 
         balance[msg.sender] = 0;
+        require(balance[msg.sender] == 0, "Failed to zero out balance");
 
-        emit TipWithdrawal(msg.sender, tipEarned);
```
Comment (farrellh1) [inline contract/contracts/Tips.sol:51]: i just watch reentrancy attack, i think we should update the balance to 0 in above before sending eth

Comment (hermannleboss) [reply contract/contracts/Tips.sol:51]: Yes, but we have reentrency gard so.

**Pull request closed without merging**